### PR TITLE
prov/tcp: rework the data code paths

### PR DESCRIPTION
**Pull request by shefty**
This series simplifies how the send and receive states are managed.  Each endpoint is active on a single transmit or receive operation at a time.  Until that operation completes, the state associated with the message does not change, except for the remaining data needing to be sent/received.  These changes target a problem where an older received header appears to be processed multiple times as a newly received message.  

The series has gone through a light set of local testing (single node in loopback) only.

### CODE REVIEW

**Comment by shefty** [comment]
The AWS failures were identified in https://github.com/ofiwg/libfabric/pull/6711.  However, I want to confirm the changes on AWS, since we've hit tcp failures in that environment that are not seen in others.  So, delaying merging until the CI there has been updated.

File: src/common.c [hunk 1]
```diff
@@ -1025,15 +1028,22 @@ size_t ofi_byteq_readv(struct ofi_byteq *byteq, struct iovec *iov,
...
 void ofi_byteq_writev(struct ofi_byteq *byteq, const struct iovec *iov,
-		      size_t cnt, size_t offset)
+		      size_t cnt)
 {
-	size_t len;
+	size_t i;
+
+	assert(ofi_total_iov_len(iov, cnt) <= ofi_byteq_writeable(byteq));
+
+	if (cnt == 1) {
```
Comment (ooststep) [inline src/common.c:1037]: I think at this level of abstraction, it isn't a problem, but I was thinking about if this optimization could affect ordering by skipping data that may be sitting in the byteq

**Comment by shefty** [comment]
uhm... I was running tests on the Intel CI nodes when the CI ran -- restarting.

File: prov/tcp/src/tcpx.h [hunk 2]
```diff
@@ -246,6 +246,7 @@ struct tcpx_fabric {
 	struct util_fabric	util_fabric;
 };
 
+#define TCXP_INTERNAL_XFER	BIT_ULL(60)
```
Comment (ooststep) [inline prov/tcp/src/tcpx.h:254]: TXPC?

Comment (shefty) [reply prov/tcp/src/tcpx.h:254]: gah -- yes -- I'm sooooooo over typing tcpx_ at this point.  It's an extreme finger exercise.  Will update

File: prov/tcp/src/tcpx_progress.c [hunk 3]
```diff
@@ -562,74 +562,63 @@ int tcpx_op_read_rsp(struct tcpx_ep *tcpx_ep)
...
-			assert(ep->cur_rx.handler);
-			ep->cur_rx.handler(ep);
+		if (ep->cur_rx.hdr_done < ep->cur_rx.hdr_len) {
+			ret = tcpx_recv_hdr(ep);
+		} else {
+			ret = ep->cur_rx.handler(ep);
 		}
 
-	} while (ofi_bsock_readable(&ep->bsock));
+	} while (!ret && ofi_bsock_readable(&ep->bsock));
```
Comment (ooststep) [inline prov/tcp/src/tcpx_progress.c:620]: should we check if the bsock is readable prior to the first recv_hdr?

File: prov/tcp/src/tcpx.h (hunk 2 above)
Comment (ooststep) [reply prov/tcp/src/tcpx.h:254]: actually, in my comment... I typed it wrong too... hah. had to edit :P

**Review by ooststep — APPROVED** [review]

File: prov/tcp/src/tcpx_progress.c (hunk 3 above)
Comment (shefty) [reply prov/tcp/src/tcpx_progress.c:620]: I don't understand the comment here.  Note that internally, all calls to receive data will always check the bsock before trying to read from the socket.  The check here is to process all prefetched data prior to exiting the progress loop, versus only handling a single message per iteration.  That should provide some benefit for small message bandwidth.

**Pull request merged**